feat: Block API

vmx opened the pull request:
This commit adds a new Block API. Encoding/decoding only happens when
it is needed. Once such an expensive operation was done, the result
is cached, so that subsequent calls can just return the result.

---

This PR isn't meant as "this should be merged exactly like this". This is meant as a starting point for further discussions. So far when I was talking about the Block API (https://github.com/ipfs-rust/rust-ipld/issues/14), things were pretty abstract, I hope that with this PR things become more concrete. This should illustrate what I'm after. Though getting there could also look completely different.

What I'd like to see that those `Block`s become a integral part, which then e.g. also the storage traits are using.

I know that @dvc94ch use cases are different from mine, so I hope we can find
solutions to both our use cases.

---

One more specific note, I'm not sure if the caching should be implemented for `BlockRef`, it might be nicer/work good enough without it.

vmx committed:
feat: Block API

This commit adds a new Block API. Encoding/decoding only happens when
it is needed. Once such an expensive operation was done, the result
is cached, so that subsequent calls can just return the result.

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+        if let Some(node) = &self.node {
+            Ok(node.clone())
+        } else if let Some(raw) = &self.raw {
+            //let decoded = Box::<dyn DynIpldCodec>::from(Box::new(self.codec))._decode(&raw).unwrap();
```
please remove unnecessary comment

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+        if let Some(raw) = &self.raw {
+            Ok(raw.clone())
+        } else if let Some(node) = &self.node {
+            //let encoded = Box::<dyn Codec>::from(self.codec)
```
same here

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+            //    .encode(&node)
+            //    .unwrap()
+            //    .to_vec();
+            let encoded = self.codec.encode(&node).unwrap().to_vec();
```
why is it safe to unwrap here? I don't think it is

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+            //    .to_vec();
+            let encoded = self.codec.encode(&node).unwrap().to_vec();
+            self.raw = Some(encoded.clone());
+            Ok(encoded)
```
should return a reference and have the caller clone it if necessary

`Ok(self.raw.as_ref().unwrap())`

vmx replied:
Thanks for mentioning the `as_ref()` stuff, that saved me a lot of time figuring things out.

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+        if let Some(cid) = &self.cid {
+            Ok(cid.clone())
+        } else {
+            // TODO vmx 2020-01-31: should probably be `encodeUnsafe()`
```
what's encodeUnsafe?

mikeal replied:
In the JS API there is `encode/decode` and `encodeUnsafe/decodeUnsafe` methods because both the binary and parsed representations are cached indefinitely and the `Unsafe` methods return those internal cached objects. In JS, we don’t have good immutability primitives, so all we can do is warn people that if they want to use these faster methods they need to make sure they don’t mutate the return values.

Rust has considerably better immutability features so I’d hope that you could find a way to achieve the same results without the `Unsafe` methods.

If I had better immutable features in JS the Block API would have returned immutable objects in both `encode` and `decode` and if someone wanted to mutate them they’d need to copy them first.

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+            Ok(cid.clone())
+        } else {
+            // TODO vmx 2020-01-31: should probably be `encodeUnsafe()`
+            let hash = Box::<dyn MultihashDigest<H>>::from(self.hash_alg).digest(&self.encode()?);
```
unnecessary boxing

vmx replied:
Hmm, I don't see how I can prevent that. Can you please help?

dvc94ch replied:
I guess it's not really possible now. You'd need to define a trait with the method `digest`, bound H to implement digest, as I've mentioned before.

vmx replied:
The trait for Multihash: https://github.com/multiformats/rust-multihash/pull/68

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,359 @@
+        // The data needs to be decoded
+        else if let Some(raw) = &self.raw {
+            let decoded = self.codec.decode(&raw).unwrap();
+            self.node_cached = Some(decoded.clone());
```
same here caller should clone

dvc94ch reviewed (commented):
There's no way of supporting native types now

vmx committed:
chore: remove left-over comments

vmx committed:
fix: return references instead of cloning

vmx committed:
fix: handle errors properly

Don't just unwrap, but transform errors.

vmx committed:
fix: use boxed slice instead of vector

Match the rest of the code base and use a boxed slice for unmutable
byte data.

vmx committed:
fix: fill in missing comments

vmx reviewed (commented):
Returning references was a great idea. I think I've addressed everything except for the Multihashing boxing (as I don't know how to do that) and we also need to look into how to support the conversion of native types.

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,368 @@
-    H: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    Box<dyn MultihashDigest<H>>: From<H>,
+    C: Copy + TryFrom<u64> + Into<u64> + EncodeDecodeIpld<H>,
+    H: Copy + TryFrom<u64> + Into<u64>,
```
Note the missing traitbound similar to `EncodeDecodeIpld<H>`

vmx committed:
fix: use DigestFromCode trait

With using the `DigestFromCode` trait we don't need to use a boxed
hasher instance.

vmx committed:
refactor: remove no longer used errors

vmx committed:
feat: re-introduce encode/decode for native types

It's again possible to directly encode native types into an IPLD `Block`.
Directlt decoding encoded data into native types is also supported.

dvc94ch commented on src/block.rs:
```diff
@@ -1,101 +1,428 @@
+            self.node = Some(decoded);
+            Ok(self.node.as_ref().unwrap())
+        } else {
+            Err(BlockError::DecodeError(
```
These are irrecoverable programmer errors. I think an immutable api could be built to make sure these can't happen at compile time.

dvc94ch reviewed (commented):
Overall looks good to me for now, but I do think that you could just as well panic.

vmx committed:
fix: declare unreachable code as such

Those cases previously returned an error, which isn't correct as those
cases cannot be reached, hence should panic if they are reaches.
